### ISSUE TITLE
Lessons learned

### ISSUE THREAD

**Issue by hgducharme** [hgducharme/meatball#3]

**Comment by hgducharme** [comment]
`CPPFLAGS` are flags that are passed to the **C** **P**re**P**rocessor
`CXXFLAGS` are flags that are passed to the C++ compiler

**Comment by hgducharme** [comment]
Prefer out-of-source builds for compiled projects. It's so much easier when using version control and keeping build artifacts separate from source files

**Comment by hgducharme** [comment]
## Compiling and Linking

Consider the directory tree

```
.
├── Makefile
├── bin
│   ├── project.out
│   └── tests.out
├── build
│   ├── src
│   │   ├── class1.o
│   │   └── main.o
│   └── test
│       └── test_class1.o
├── src
│   ├── class1.cpp
│   ├── class1.h
│   └── main.cpp
└── test
    └── test_class1.cpp
```

**Suppose:**
- `main.cpp` includes `class1.h`
- `test_class1.cpp` includes `class1.h`

## Making the `project.out` executable

**Working backwards:**
1. In order to create the `bin/project.out` executable, we have to link together the `build/src/class1.o` and `build/src/main.o` files.
2. To create the `class1.o` file we have to compile the `class1.cpp` file
3. To create the `main.o` file we have to compile the `main.cpp` file

```
# Compile class1.cpp
clang++ --compile src/class1.cpp --output build/src/class1.o

# Compile main.cpp
clang++ --compile src/main.cpp --output build/src/main.o

# Link them together to produce an executable
clang++ build/src/class1.o build/src/main.o --output bin/project.out
```

You can compile implicitly like

```
clang++ /src/main.cpp /src/class1.cpp --output /bin/project.out
```

but then you don't get .o files. Not quite sure the cons of this approach.

## Making the `tests.out` executable

**Working backwards:**
1. In order to create the `bin/tests.out` executable, we have to link together the `build/src/class1.o` and `build/test/test_class1.o` files.
2. To create the `class1.o` file we have to compile the `class1.cpp` file
3. To create the `test_class1.o` file we have to compile the `test_class1.cpp` file

```
# Compile class1.cpp
clang++ --compile src/class1.cpp --output build/src/class1.o

# Compile test_class1.cpp
clang++ --compile test/test_class1.cpp --output build/test/test_class1.o

# Link them together to produce an executable
clang++ build/src/class1.o build/test/test_class1.o --output bin/tests.out
```

**Comment by hgducharme** [comment]
When dealing with binary data and bit patterns in C/C++, people will generally represent these with hexadecimal (e.g. "0xffff0000") or octal numbers. There's a direct mapping between hex/octal numbers and bit patterns that makes the code a bit more readable and understand, which isn't available when you use decimal values (e.g. "4571894638") to represent bit patterns.

Binary literals weren't around in C until C23 and C++ until C++14. We have access to binary literals no but this is why you will see. To see how to use binary literals see [this](https://www.geeksforgeeks.org/binary-literals-in-c14-with-examples/#).

Tangentially related, a bitmask is just a value that is used with another value to perform certain operations on the bits of the second value. Think of it like a linear operator in linear algebra: you apply the linear operator `A` to a vector `x` in order to perform some operation on that matrix and get the resulting vector `b` (`Ax = b`).

**Comment by hgducharme** [comment]
See [cppreference on binary operators](https://en.cppreference.com/w/cpp/language/operators) for best practices on overloading bitwise operators

**Comment by hgducharme** [comment]
If a function definition returns a reference like

```
X & function()
```

but you don't specify a reference when calling it, you will not get a reference to the original value. You will copy it maybe?

```cpp
// Will store the value in a NEW value called var. Any modifications to var will not modify the original value
X var = function();

// Will obviously return a reference to the original value. Any modification to var will modify the original value
X & var = function();
```

**Comment by hgducharme** [comment]
https://stackoverflow.com/questions/16586489/c-interface-vs-template

**Comment by hgducharme** [comment]
## Bitboard concerns

I'm starting to question my use of my `Bitboard` class that acts as a wrapper for the primitive type `uint64_t`. I like the syntatic sugar it provides such as

```
Bitboard blackPawns(DEFAULT_BLACK_PAWN_STRUCTURE);
blackPawns.setBit(3);
blackPawns.numberOfSetBits();
blackPawns.findIndexLSB();
```

instead of

```
u64 blackPawns = DEFAULT_BLACK_PAWN_STRUCTURE)
setBit(blackPawns, 3);
numberOfSetBits(blackPawns);
findIndexLSB(blackPawns);
```

## Concern

I'm concerned about future performance. I'm starting to see that I need to instantiate the class just to represent something as a `uint64_t`and pass that to something else. I feel like this is unnecessarily using a lot of stack memory which will catch up to me later on by causing speed issues in the engine.

**Comment by hgducharme** [comment]
- Strategy pattern: when you need to swap out behavior implementations during _runtime_ (e.g. the payment system needs to pay with paypal, venmo, card, or cash during runtime)
- Template pattern: when you have an algorithm with invariant parts and need to slightly modify parts of the algorithm depending on the strategy being used (e.g. you have a mostly invariant algorithm for setting up a game character, but some characters need a little additional initialization during the creation process)
- Template meta-programming: when an algorithm will work with different variable types, use templates

Side note:
- I liked [this answer's](https://stackoverflow.com/a/772727) part about reminding oneself to not get carried away.

**Comment by hgducharme** [comment]
## Magic bitboards

### Process

1. Generate a bitmask of the squares that could potentially be blocking a slider piece. Commonly called the relevant occupancy bits, or the relevant occupancy set. For example, for a rook on A1, the relevant occupancy set would be the squares A2-A7 and B1-G1. They is the key to the hash map.
2. Multiply the relevant occupancy bitmask by a "magic number" (see below for finding magic numbers). This is the index used for lookup later on.
3. Right shift the resulting product by `64 - n` bits, where `n` is the number of bits in the resulting product `i`.
4. Use the index `i` generated in step 3 to access a pre-initialized move database

### What are "magic numbers"

The goal of a magic number is such that when it is multiplied against an occupancy mask, it maps the scattered placement of occupied bits into a consecutive ordering in the most significant bit positions. The equation is:
```
index = (occupancy set * magic number) >> (64 - number of bits in occupancy set)
```
For example, for a bishop on B1:

```
                                          any consecutive
relevant occupancy                        combination of
for bishop on b1,     magic number        the masked bits
5 bits
. . . . . . . .      . . . . . . . .     . . .[C D E F G]
. . . . . . . .      . 1 . . . . . .     . . . . . . . .
. . . . . . G .      . 1 . . . . . .     . . . . . . . .
. . . . . F . .      . 1 . . . . . .     . . . . . . . .
. . . . E . . .   *  . 1 . . . . . .  =  . . garbage . .   >> (64- 5)
. . . D . . . .      . 1 . . . . . .     . . . . . . . .
. . C . . . . .      . . . . . . . .     . . . . . . . .
. . . . . . . .      . . . . . . . .     . . . . . . . .
```

### Finding magic numbers

**Comment by hgducharme** [comment]
**Question**: What is the performance difference between `k` for loops all with simple instructions versus one for loop with all instructions inside it, and all for loops have the same amount of iteration loops?

**Research**
- [Stack exchange](https://gamedev.stackexchange.com/questions/157001/why-do-separate-loops-run-faster-than-one-loop-that-does-multiple-things)
- [Stack exchange 2](https://stackoverflow.com/questions/62606991/using-one-loop-vs-two-loops)

**Consensus**
Apparently `k` multiples of a for loop can theoretically be faster than one for loop with a lot of complex instructions. Both situations are obviously `O(n)`, but there might be slight optimizations that can be made. I suppose it really depends on the context and this is kind of considered a micro optimization, which shouldn't be done unless you profiled the code and have evidence that one is a bottleneck compared to the other.

**Decision**
I personally sort of like the idea of multiple for loops with simple instructions in them because it allows me to write functions with good encapsulation and naming, and therefore aiding in readability and maintainability. I suppose I will go ahead with that route and can profile the code if this is slow and I need it to be faster.

**Comment by hgducharme** [comment]
Apparently you can't initialize C++ arrays (in shorthand) with a value other than zero. That is, the following would not behave as you expect

```cpp
int myArray[10][10] = { {50} };
```

The first element would be 50 but all other elements would be zero. You can use `std::fill` to set all elements to 50. For a concise answer on why, see [this stackoverflow answer](https://stackoverflow.com/a/15521368).

**Comment by hgducharme** [comment]
[Cool video] on psuedo random number generation in c++ and why `rand()` is harmful

**Comment by hgducharme** [comment]
`constexpr` functions require a [literal](https://en.cppreference.com/w/cpp/named_req/LiteralType) return type.

**Comment by hgducharme** [comment]
- Profiling: analyzing a program to see how long certain parts of the program take to execute. The result of a code profile might be a table of the functions that are being executed and how much time each function took to execute
- Benchmarking:  analyzes how fast some operation takes to run. The result of a benchmark should be a singular value
- Microbenchmarking: measuring the performance of one super specific task, in isolation from anything else.

https://stackoverflow.com/questions/34801622/difference-between-benchmarking-and-profiling

**Comment by hgducharme** [comment]
## `Static` keyword

**On class member functions and variables**: defining a class member function or variable as `static` means it belongs to the class and not any particular instance of the class

**On non-member functions and variables**: defining a non-member function or variable as `static` means it only has internal linkage and will not be used outside of its translation unit. This use of `static` has been replaced by anonymous namespaces, but they both do exactly the same thing. The benefit of using `static` is that it's clear by function definition alone that it has internal linkage, and there's no need to scroll up or down to see if it's contained in an anonymous namespace.

**Comment by hgducharme** [comment]
Have learned a lot about docker implementing this CI pipeline. There's some super niche quirks when integrating with Github Actions.

One thing that was tripping me up was that I was defining a non-root user in my docker containers but this doesn't work on Github Action servers, as detailed [in this article](https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions).

**Comment by hgducharme** [comment]
### Namespaces

Ok wow just had to detangle the biggest mess with an unnamed namespace nested within a named namespace and split between the interface file `.h` and the implementation file `.cpp`

What I had before:

```
// .h

namespace xyz {
   void a();
   
   namespace {
      void b();
      void c();
   }
}

// .cpp

namespace xyz {
   void a() { b(); }
   
   namespace {
      void b() { c(); }
      void c() { ... }
   }
}
```

Technically this will compile and work just fine but you will get the error in the header file for `b()` saying it was declared static but never defined, and therefore it's an unused function. 

Ultimately you're not supposed to put anonymous namespaces in header files, so I went ahead and removed the anonymous namespace from the header file, but then in the source file you will get the error that `b()` is not defined in the scope of `a()`, and therefore `a()` can't call `b()`. The issue here is that now our function definitions simultaneously become declarations also, and since c++ presumably parses the file from the top down, then it hasn't seen the declaration for `b()` when it parses `a()`. Therefore, you need to move the anonymous namespace to the top of the named namespace. However, you will still have the function ordering problem within the unnamed namespace. You must also keep the unnamed namespace inside the named namespace for `a()` to see `b()`.

**TLDR: The following is what works**

```
// .h

namespace xyz {
   void a();
}

// .cpp

namespace xyz {

   namespace {
      void c() { ... }
      void b() { c(); }
   }
   
   void a() { b(); }
}
```

**Comment by hgducharme** [comment]
Cool discussions on array of structs vs array of pointers to structs:
- [1](https://stackoverflow.com/questions/42328735/array-of-structs-vs-array-of-pointers-to-structs)
- [2](https://stackoverflow.com/questions/6541387/use-an-array-of-pointers-to-structs-or-just-an-array-of-structs?rq=3)

Ran into this question when considering how I should manage `BISHOP_HASHING_PARAMETERS_LOOKUP` and `ROOK_HASHING_PARAMETERS_LOOKUP`. Right now they are arrays of structs, and I will just keep it like that until I profile the code and see where bottlenecks are

**Comment by hgducharme** [comment]
## Interfaces vs Abstract Base Classes (ABC) in C++

- _Abstract base class_ (ABC; also called abstract class): a c++ class that contains a pure virtual function
- _Interface_ (also called a pure abstract class): a c++ class that contains only pure virtual functions **and** a virtual destructor (doesn't have to be pure virtual)
- _Virtual function_: a member function that declared in a base class that can be overridden
- _Pure virtual function_: a member function declared in a base class that has no definition, and **must** be overridden by a child class

```cpp
class Abstract
{
   Abstract();
   ~Abstract();
   void functionWithADefinition { does stuff... };
   virtual void virtualFunctionWithDefinition { does stuff... };
   virtual void pureVirtual()  = 0;

   // Abstract classes can contain member variables
   int memberVariable = 0;
}

class Interface
{
   virtual ~Interface() = 0; // or virtual ~Interface() {};
   virtual void pureVirtual1 = 0;
   virtual int pureVirtual2()  = 0;
}
```

Both abstract classes and interfaces define the concept of an interface (or API) to clients that use them, the only difference is that abstract classes have pre-packaged functionality that is shared between implementations. Use an interface if there is no shared functionality. Use an abstract class if there isn't one.

All interfaces (pure abstract classes) are abstract classes, but not all abstract classes are interfaces. Before using either, make sure the implementations has an _is-a_ relationship with the parent class.

**Comment by hgducharme** [comment]
When this is ultimately turned into a technical presentation, the lessons learned should be formatted by categories:

1. System architecture lessons (e.g. how the different modules of the system interact with each other)
2. System design lessons (e.g. the design patterns used within just one module)
3. Programming lessons (e.g. compilation, computer science, C++ specific things)
5. Chess specific
